internal/kafka/internal/handlers/validation.go

```diff
@@ -407,3 +407,51 @@ func validateKafkaMachinePoolNodeCount(clusterPayload *public.EnterpriseOsdClust
 		return nil
 	}
 }
+
+func validateEnterpriseClusterEligibleForDeregistration(ctx context.Context, clusterID string, forceDeleteFlag bool, clusterService services.ClusterService) handlers.Validate {
+	return func() *errors.ServiceError {
+		claims, claimsErr := getClaims(ctx)
+
+		if claimsErr != nil {
+			return claimsErr
+		}
+
+		orgID, getOrgIdErr := claims.GetOrgId()
+
+		if getOrgIdErr != nil {
+			return errors.GeneralError(getOrgIdErr.Error())
+		}
+
+		if !claims.IsOrgAdmin() {
+			return errors.New(errors.ErrorUnauthorized, "non admin user not authorized to perform this action")
+		}
+
+		cluster, err := clusterService.FindClusterByID(clusterID)
+		if err != nil {
+			return err
+		}
+
+		if cluster != nil {
+			if cluster.OrganizationID != orgID {
+				return errors.Forbidden("unable to deregister cluster from different organization")
+			}
+			if cluster.ClusterType != api.EnterpriseDataPlaneClusterType.String() {
+				return errors.Forbidden("unable to deregister cluster whose type is not: %s", api.EnterpriseDataPlaneClusterType.String())
+			}
+			if !forceDeleteFlag {
+				clusterKafkaCount, err := clusterService.FindKafkaInstanceCount([]string{clusterID})
+				if err != nil {
+					return errors.GeneralError(err.Error())
+				}
+				for _, count := range clusterKafkaCount {
+					if count.Clusterid == clusterID && count.Count > 0 {
+						return errors.Forbidden("unable to deregister %s cluster with id: %s without explicitly setting 'force:true' due to kafka requests present on this cluster",
+							api.EnterpriseDataPlaneClusterType.String(), clusterID)
+					}
+				}
```

Comment:
The loop is not required. We could just check if `len(clusterKafkaCount) > 0` and return the error returned in line 448.

Reply:
but clusterKafkaCount might be greater than 0 and don't contain entry for the cluster in question

Reply:
From here https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/8cb312233571ae936ea9643379969fb30642634c/internal/kafka/internal/handlers/validation.go#L442 we are only filtering for the given cluster. The output is either empty if the cluster in question does not contain any kafka, otherwise it returns a list of kafkas of the given cluster and nothing else.

Reply:
good point. changed

Reply:
@pawelpaszki I noticed that the CI was failing due to my suggestion. I was mistaken https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1473#discussion_r1064574003 and my statement below 

> The output is either empty if the cluster in question does not contain any kafka, otherwise it returns a list of kafkas of the given cluster and nothing else.

is incorrect and misleading. 

The correct behaviour is:
1. if the cluster in question contains Kafka, a non empty list is returned. The list has one element whose Count reflects the total streaming unit count.
2. if the cluster in question doesn't contain any kafka, a non empty list is returned. The list has one element whose Count is `0`.

Reply:
Apologies for the wrong suggestion earlier on.
 
We could either keep the loop knowing that the list will always contain one element.
Or we could just fetch the first item of the list and work with it knowing that the list will always contain one element.
we should be guarded from future regressions by the tests.